### PR TITLE
Mm 68768 auto cache rolling cleanup

### PR DESCRIPTION
**Pull request by carlisgg**
#### Summary
Auto cache cleanup for posts, playbook runs, and AI threads.

Adds automatic rolling cleanup of cached posts, playbook runs, and AI agent threads so local data doesn't grow unbounded, centralized in a new `app/actions/local/ephemeral_mode/cleanup.ts`

Most post cleanup processing lives in the SQL/query layer (`app/queries/servers/post.ts`) to make the cleanup atomic, while playbook runs and AI threads processing lives in the JS-side due to its simplicity.

Post ranges kept in POSTS_IN_CHANNEL rows are reconciled via watermelonDB operator model handling. This method is further protecting mounted channel re-firing observers and letting the UI properly refresh to keep using it without having inconsistencies in cached and fetched data.

Cleanup process runs daily and there is an inflight marker preventing overlapping runs for the same server.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-68768

#### Checklist
- [x] Added or updated unit tests (required for all new features)

#### Device Information
This PR was tested on:
- Android 17 and iOS 26.5 emulators
- Pixel 6A with Android 17
- iPhone 15 with iOS 26.5

#### Release Note
```release-note
- Adds support for auto cache cleanup setting coming from server and preventing CUI accumulation by performing rolling cleanups on cached content.
```

### CODE REVIEW

File: app/actions/local/post.ts [hunk 1]
```diff
@@ -414,6 +414,57 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+                    [`DELETE FROM ${THREAD_PARTICIPANT} WHERE thread_id IN (${postSubquery})`, []],
+                    [`DELETE FROM ${THREADS_IN_TEAM} WHERE thread_id IN (${postSubquery})`, []],
+                    [`DELETE FROM ${POST} WHERE ${postCondition}`, []],
+
+                    // PostsInChannel bookkeeping, delete or update values depending on posts left in the channel after the cutoff
+                    [`DELETE FROM ${POSTS_IN_CHANNEL} WHERE channel_id IN (${channels}) AND latest < ${cutoff}`, []],
+                    [`UPDATE ${POSTS_IN_CHANNEL} SET earliest = ${cutoff} WHERE channel_id IN (${channels}) AND earliest < ${cutoff}`, []],
+
+                    // MyChannel bookkeeping, reset last_fetched_at for channels that have no posts left after the cutoff
+                    [`UPDATE ${MY_CHANNEL} SET last_fetched_at = 0 WHERE id IN (${channels}) AND last_fetched_at > 0 AND NOT EXISTS (SELECT 1 FROM ${POSTS_IN_CHANNEL} WHERE channel_id = ${MY_CHANNEL}.id)`, []],
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/post.ts:466]: we cannot extend the range to a single protected post. It will self heal on next fetch.

File: app/managers/ephemeral_mode_manager/index.ts [hunk 2]
```diff
@@ -136,11 +145,18 @@ class EphemeralModeManagerSingleton {
...
         timeoutStr: string | undefined,
         purgeHoursStr: string | undefined,
+        cleanupDaysStr: string | undefined,
     ) => {
         const nextEnabled = enabledStr === 'true';
         const nextThresholdMs = Math.max(0, Number(timeoutStr ?? '0')) * 1000;
         const nextPurgeThresholdMs = Math.max(0, Number(purgeHoursStr ?? '0')) * 3600 * 1000;
+        const nextCleanupDays = Math.max(0, Number(cleanupDaysStr ?? '0'));
         const wasActive = this.trackedServers.get(serverUrl)?.kind === 'mem';
+        this.cleanupDays[serverUrl] = nextCleanupDays;
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/managers/ephemeral_mode_manager/index.ts:155]: server ensures a numeric value is received

File: app/actions/local/ephemeral_mode/cleanup.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        await operator.batchRecords(prepared, 'cleanupPlaybookRuns');
+    }
+}
+
+export async function autoCacheCleanup(serverUrl: string): Promise<void> {
+    const cleanupDays = EphemeralModeManager.getAutoCacheCleanupDays(serverUrl);
+    if (cleanupDays <= 0) {
+        logDebug('autoCacheCleanup: cleanupDays <= 0 for', serverUrl);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.ts:228]: it is numeric

File: app/screens/pdf_viewer/pdf_viewer.tsx [hunk 4]
```diff
@@ -134,6 +137,26 @@ const PdfViewer = ({allowPdfLinkNavigation, fileId, filePath, siteURL}: Props) =
...
+            try {
+                const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
+                const file = await getFileById(database, fileId);
+                if (active && file?.postId) {
+                    EphemeralStore.setCurrentFileViewerPostId(file.postId);
+                }
+            } catch {
+                // file not in DB — no protection needed
+            }
+        })();
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/screens/pdf_viewer/pdf_viewer.tsx:153]: no protectin is needed, logging the error is not important

File: app/actions/websocket/index.ts [hunk 5]
```diff
@@ -112,6 +113,8 @@ async function doReconnect(serverUrl: string, groupLabel?: BaseRequestGroupLabel
 
         expiredBoRPostCleanup(serverUrl);
 
+        autoCacheCleanup(serverUrl);
+
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/websocket/index.ts:117]: that is not the case

File: app/actions/local/ephemeral_mode/cleanup.ts [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    let threadParentLimit = Infinity;
+
+    if (viewedThreadId) {
+        activeThreadRootIds.add(viewedThreadId);
+        const rootPost = await getPostById(database, viewedThreadId);
+        if (rootPost) {
+            threadParentChannelId = rootPost.channelId;
+
+            // Floored at the root's own create_at so the open thread's replies (always >= root.createAt)
+            // are never deleted, even when there aren't enough preceding posts to compute a wider buffer.
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.ts:101]: floored os fine

File: app/queries/servers/post.ts [hunk 7]
```diff
@@ -427,3 +427,38 @@ export const syncPermalinkPreviewsForEditedPost = async (
...
+// These are live threads — their root post must not be evicted even if its
+// create_at is below the cutoff.
+export const queryActiveThreadRootIds = async (
+    database: Database,
+    cutoff: number,
+): Promise<Set<string>> => {
+    const rows = await database.get<PostsInThreadModel>(POSTS_IN_THREAD).query(
+        Q.where('latest', Q.gte(cutoff)),
+    ).fetch();
+    return new Set(rows.map((r) => r.rootId));
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/queries/servers/post.ts]: this is about replies not the root id itself, so threads with replies more recent than cutoff (latest in the PiT) should survive regardless of earliest.

File: app/managers/ephemeral_mode_manager/index.ts [hunk 8]
```diff
@@ -136,11 +145,18 @@ class EphemeralModeManagerSingleton {
...
+
+        if (nextCleanupDays > 0) {
+            logDebug('EphemeralModeManager: auto cache cleanup config received, days:', nextCleanupDays, 'for', serverUrl);
+        }
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/managers/ephemeral_mode_manager/index.ts:159]: serverURL is already used in some other logDebug cases

File: app/actions/local/ephemeral_mode/cleanup.ts [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+import {queryPlaybookRunsBefore} from '@playbooks/database/queries/run';
+import {
+    createAtOfNthPostOlderThan,
+    getPostById,
+} from '@queries/servers/post';
+import {getCurrentChannelId} from '@queries/servers/system';
+import EphemeralStore from '@store/ephemeral_store';
+import {NavigationStore} from '@store/navigation_store';
+import {getFullErrorMessage} from '@utils/errors';
+import {logDebug, logError} from '@utils/log';
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.ts:21]: serverURL is already logged in many other places

File: app/actions/local/post.ts [hunk 10]
```diff
@@ -414,6 +416,78 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
+
+        const channels = `'${channelIds.join("','")}'`;
+        const exclusionClause = excludedPostIds.size > 0 ? ` AND id NOT IN ('${[...excludedPostIds].join("','")}')` : '';
+
+        // A thread with a reply on or after the cutoff is alive even if its root is older;
+        // keep the root so the surviving reply's root_id never dangles.
+        const hasActiveReply = `EXISTS (SELECT 1 FROM ${POSTS_IN_THREAD} WHERE ${POSTS_IN_THREAD}.root_id = ${POST}.id AND ${POSTS_IN_THREAD}.latest >= ${cutoff})`;
+
+        const postCondition = `channel_id IN (${channels}) AND create_at < ${cutoff} AND NOT ${hasActiveReply}${exclusionClause}`;
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/post.ts:440]: Not a defect. A reply surviving without its local root isn't a state the cleanup uniquely creates, it's already routine. This BTW only matters if the file was opened without first visiting a channel or thread post list.

File: app/actions/local/post.ts [hunk 11]
```diff
@@ -414,6 +416,78 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+    excludedPostIds: Set<string> = new Set(),
+    reconcileObservers = false,
+): Promise<{error: unknown}> {
+    if (channelIds.length === 0) {
+        return {error: undefined};
+    }
+
+    try {
+        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
+
```
Comment (edgarbellot) [inline app/actions/local/post.ts:432]: The 11 statements here build their `WHERE` clauses by string concatenation and always pass an empty args array, so `channelIds` and `excludedPostIds` go straight into the SQL text. Those ids come from server JSON and nothing validates the 26-char format, so an id containing a single quote would change the shape of the query and make the DELETE unbounded.

The batch path does support bound args (`[cacheBehavior, table, sql, [args]]`), so could we bind the id lists with `?` placeholders instead? Failing that, an id-format filter before building the strings would do. Worth adding a test with a quote-containing id too, since the current assertions in post.test.ts just compare the interpolated SQL string.

File: app/managers/ephemeral_mode_manager/index.ts [hunk 12]
```diff
@@ -136,11 +145,18 @@ class EphemeralModeManagerSingleton {
         enabledStr: string | undefined,
         timeoutStr: string | undefined,
         purgeHoursStr: string | undefined,
+        cleanupDaysStr: string | undefined,
     ) => {
         const nextEnabled = enabledStr === 'true';
         const nextThresholdMs = Math.max(0, Number(timeoutStr ?? '0')) * 1000;
         const nextPurgeThresholdMs = Math.max(0, Number(purgeHoursStr ?? '0')) * 3600 * 1000;
+        const nextCleanupDays = Math.max(0, Number(cleanupDaysStr ?? '0'));
```
Comment (edgarbellot) [inline app/managers/ephemeral_mode_manager/index.ts:153]: Two things about this value.

`Math.max(0, Number('abc'))` is `NaN`, and `NaN <= 0` is false, so a non-numeric config value slips past the guard in cleanup.ts:225 and ends up in the SQL as `create_at < NaN`. The transaction fails as a whole so there is no data loss, but `LAST_AUTO_CACHE_CLEANUP_RUN` is only written on success, so the full sweep retries on every reconnect. Could we add a `Number.isFinite` check plus a minimum here?

Separately, `this.cleanupDays[serverUrl]` is assigned regardless of `nextEnabled`, and the server emits `AutoCacheCleanupDays` behind the feature flag but outside the `ephemeralEnabled` check (client.go:263-273). So a server with the flag on and ephemeral mode off still gets the 7-day rolling purge on every client. Is that intended, or should `autoCacheCleanup` be gated on `MobileEphemeralModeEnabled`?

File: app/actions/local/post.ts [hunk 13]
```diff
@@ -414,6 +416,78 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+        const hasActiveReply = `EXISTS (SELECT 1 FROM ${POSTS_IN_THREAD} WHERE ${POSTS_IN_THREAD}.root_id = ${POST}.id AND ${POSTS_IN_THREAD}.latest >= ${cutoff})`;
+
+        const postCondition = `channel_id IN (${channels}) AND create_at < ${cutoff} AND NOT ${hasActiveReply}${exclusionClause}`;
+        const postSubquery = `SELECT id FROM ${POST} WHERE ${postCondition}`;
+
+        // Scopes PostsInThread trimming to roots in these channels.
+        const rootInChannelsExists = `EXISTS (SELECT 1 FROM ${POST} WHERE ${POST}.id = ${POSTS_IN_THREAD}.root_id AND ${POST}.channel_id IN (${channels}))`;
+
+        await database.write(() => {
+            return database.adapter.unsafeExecute({
```
Comment (edgarbellot) [inline app/actions/local/post.ts:447]: `unsafeExecute` maps every statement with `cacheBehavior = IGNORE_CACHE` and bypasses `Database.batch`, so nothing invalidates `Collection._cache`. `RecordCache._modelForRaw` returns the cached record and discards the fresh raw when the id is already in the map.

That matters for the raw `PostsInChannel` earliest bump on line 463. Any chunk already loaded in memory keeps its pre-cleanup `earliest`, and `handleReceivedPostsInChannel` re-reads chunks via `query().fetch()` and then does `record.earliest = Math.min(record.earliest, earliest)` (handlers/post.ts:610), writing the old value back over the trimmed row. The channel then looks like it covers a range whose posts are gone, so `queryPostsChunk` returns nothing there and pagination never re-requests them.

`reconcileObservers` already fixes this for the mounted channels. Could we run that reconciliation for all touched channels? It is one row per channel, so the batch is cheap.

File: app/actions/local/post.ts [hunk 14]
```diff
@@ -414,6 +416,78 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+                sqls: [
+                    [`DELETE FROM ${REACTION} WHERE post_id IN (${postSubquery})`, []],
+                    [`DELETE FROM ${FILE} WHERE post_id IN (${postSubquery})`, []],
```
Comment (edgarbellot) [inline app/actions/local/post.ts:450]: This deletes the `File` rows but I could not find any disk-side deletion in the cleanup path, and there is no age or size based pruning of the per-server file directory anywhere else (the only `deleteFileCache` callers are logout, advanced settings and the ephemeral-mode wipe). So downloaded attachments stay on disk with no DB row pointing at them, which makes them unreachable for any later cleanup.

Attachments are usually the biggest contributor to local growth, so this works against the goal of the PR. Could we collect the local paths before deleting the rows and remove them from disk in the same run?

Note the pre-existing `deletePosts` path has the same gap, so this is not introduced here. Raising it because this PR makes the deletion a daily behaviour on all servers.

File: app/actions/websocket/index.ts [hunk 15]
```diff
@@ -112,6 +113,8 @@ async function doReconnect(serverUrl: string, groupLabel?: BaseRequestGroupLabel
 
         expiredBoRPostCleanup(serverUrl);
 
+        autoCacheCleanup(serverUrl);
```
Comment (edgarbellot) [inline app/actions/websocket/index.ts:116]: `dataRetentionCleanup` and `autoCacheCleanup` are both fired without `await`, both are gated to run once per day, and both end with `database.unsafeVacuum()` (systems.ts:265 and cleanup.ts:286). So on the first reconnect of each day they overlap, and whichever vacuums first makes the other's remaining DB work throw. The cleanup then logs the error, does not stamp its timestamp, and retries on the next reconnect.

Should we await them in sequence, or share a single daily vacuum between the two jobs?

File: app/actions/local/ephemeral_mode/cleanup.ts [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+            );
+
+            await cleanupPosts(serverUrl, database, cutoff, limits);
+            await cleanupAiThreads(database, operator, cutoff, limits.viewedThreadId);
+            await cleanupPlaybookRuns(database, operator, cutoff, limits.viewedPlaybookRunId);
+
+            await setLastAutoCacheCleanupRun(serverUrl);
+
+            try {
+                await database.unsafeVacuum();
```
Comment (edgarbellot) [inline app/actions/local/ephemeral_mode/cleanup.ts:286]: `Database.unsafeVacuum` swaps `this.adapter` for `ErrorAdapter` for the whole operation, and unlike `unsafeResetDatabase` it does not call `_ensureInWriter` or `_abortPendingWork`. So while it runs, every `find`, `query` and `batch` in the app throws: mounted observers, websocket handlers persisting incoming posts, message sending. This happens on a reconnect with the app in the foreground.

Since `dataRetentionCleanup` already vacuums daily on the same reconnect, could we drop the vacuum here, or make it conditional on the run having actually deleted a meaningful amount?

File: app/actions/local/post.ts [hunk 17]
```diff
@@ -414,6 +416,78 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+
+                    // PostsInThread bookkeeping, delete or update values depending on replies left in the thread after the cutoff
+                    [`DELETE FROM ${POSTS_IN_THREAD} WHERE latest < ${cutoff} AND ${rootInChannelsExists}`, []],
+                    [`UPDATE ${POSTS_IN_THREAD} SET earliest = ${cutoff} WHERE earliest < ${cutoff} AND ${rootInChannelsExists}`, []],
+
+                    [`DELETE FROM ${THREAD} WHERE id IN (${postSubquery})`, []],
+                    [`DELETE FROM ${THREAD_PARTICIPANT} WHERE thread_id IN (${postSubquery})`, []],
+                    [`DELETE FROM ${THREADS_IN_TEAM} WHERE thread_id IN (${postSubquery})`, []],
+                    [`DELETE FROM ${POST} WHERE ${postCondition}`, []],
```
Comment (edgarbellot) [inline app/actions/local/post.ts:459]: Keeping drafts out of this cleanup is the right call, but it leaves reply drafts whose `root_id` points at a root post that was just deleted.

The draft still shows up in the list, since `queryDraftsForTeam` only joins on Channel and the list item renders from the draft model alone. But tapping it calls `switchToThread`, which resolves the root with `getPostById`, throws `Post not found` with no server fallback, and only logs the error (thread.ts:68-71). So the tap silently does nothing. Same for "Edit draft". Sending from the list still works, since the root exists server-side.

Could we either exclude posts that have a draft rooted on them (a `NOT EXISTS (SELECT 1 FROM Draft WHERE root_id = Post.id)` in `postCondition`), or make `switchToThread` fetch the root from the server when it is missing locally? The second option also helps permalinks and notifications pointing at evicted roots.

**Review by edgarbellot — CHANGES_REQUESTED** [review]
@carlisgg left some suggestions, let me know if you have any question :)

File: app/actions/local/post.ts (hunk 11 above)
Comment (carlisgg) [reply app/actions/local/post.ts:432]: channelIds and excludedPostIds are coming from server data and not provided by the user. Still switching to use bound args is pretty easy so will do it

File: app/actions/websocket/index.ts (hunk 15 above)
Comment (carlisgg) [reply app/actions/websocket/index.ts:116]: I think share a single vaccum and plan for it to run awaited on next reconnection is the right call. Anyway, since this is common to both jobs and an actual issue, I will open a JIRA ticket just for this

File: app/actions/local/ephemeral_mode/cleanup.ts (hunk 16 above)
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.ts:286]: same as for previous comment about vacuum and awaits

File: app/actions/local/post.ts (hunk 17 above)
Comment (carlisgg) [reply app/actions/local/post.ts]: done

File: app/actions/websocket/index.ts (hunk 15 above)
Comment (carlisgg) [reply app/actions/websocket/index.ts]: ticket open at https://mattermost.atlassian.net/browse/MM-70200

File: app/actions/local/post.ts [hunk 18]
```diff
@@ -414,6 +417,97 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
...
+                    r.earliest = cutoff;
+                })),
+            ...myChannelRows.
+                filter((row) => row.lastFetchedAt > 0 && !channelsWithSurvivingPosts.has(row.id)).
+                map((row) => row.prepareUpdate((r) => {
+                    r.lastFetchedAt = 0;
+                })),
+        ];
+        if (prepared.length > 0) {
+            await operator.batchRecords(prepared, 'deletePostsInChannelsByCutoff.reconcile');
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/post.ts:504]: watermelondb treats deleted models on a different way and no need to refire via the model layer.

File: app/managers/ephemeral_mode_manager/index.test.ts [hunk 19]
```diff
@@ -201,6 +208,19 @@ describe('EphemeralModeManager', () => {
...
+        wsStates[serverA] = new BehaviorSubject<WebsocketConnectedState>('connected');
+
+        await EphemeralModeManager.init([credsA]);
+        await advanceTimers(0);
+
+        setWs(serverA, 'not_connected');
+        await advanceTimers(0);
+
+        expect(EphemeralModeManager.isOffline(serverA)).toBe(true);
+    });
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/managers/ephemeral_mode_manager/index.test.ts:222]: following existing file convention

File: app/actions/local/post.ts (hunk 14 above)
Comment (carlisgg) [reply app/actions/local/post.ts]: created issue at https://mattermost.atlassian.net/browse/MM-70201

File: app/actions/local/post.test.ts [hunk 20]
```diff
@@ -400,3 +405,174 @@ describe('updatePostTranslation', () => {
...
+    async function writePostsInThread(rootId: string, earliest: number, latest: number): Promise<PostsInThreadModel> {
+        let record!: PostsInThreadModel;
+        await operator.database.write(async () => {
+            record = await operator.database.get<PostsInThreadModel>(POSTS_IN_THREAD).create((r) => {
+                r.rootId = rootId;
+                r.earliest = earliest;
+                r.latest = latest;
+            });
+        });
+        return record;
```
[the remark this answers is not in the document]
Comment (carlisgg) [reply app/actions/local/post.test.ts:479]: it is fine in this test file

File: app/screens/channel/channel_post_list/channel_post_list.tsx [hunk 21]
```diff
@@ -105,12 +108,30 @@ const ChannelPostList = ({
...
     useDidMount(() => {
         return () => {
+            EphemeralStore.setCurrentChannelOldestVisibleCreateAt(undefined);
             unsetActiveChannelOnServer(serverUrl);
         };
     });
 
+    const onViewablePostsChanged = useCallback((viewableItems: ViewToken[]) => {
+        let oldest = Infinity;
+        for (const {item, isViewable} of viewableItems) {
```
Comment (larkox) [inline app/screens/channel/channel_post_list/channel_post_list.tsx]: Do we know anything about the order of viewable items? I understand that there may be non-viewable items in the list (which sounds weird), and items that are not posts, so we probably have to do the loop either way. But, if I am not mistaken, the first or the last post in the list should be the oldest, so checking just one and breaking the loop should be enough.

File: app/queries/servers/post.ts
Comment (larkox) [inline app/queries/servers/post.ts:1]: Missing tests?

File: app/products/playbooks/screens/playbook_run/playbook_run.tsx [hunk 22]
```diff
@@ -165,6 +167,7 @@ const getStyleSheet = makeStyleSheetFromTheme((theme) => ({
 }));
 
 type Props = {
+    playbookRunId: string;
```
Comment (larkox) [inline app/products/playbooks/screens/playbook_run/playbook_run.tsx:210]: Why the new prop, and not directly the id from `playbookRun`?

File: app/managers/ephemeral_mode_manager/index.ts [hunk 23]
```diff
@@ -121,13 +131,16 @@ class EphemeralModeManagerSingleton {
             observeConfigValue(database, 'MobileEphemeralModeEnabled'),
             observeConfigValue(database, 'MobileEphemeralModeDisconnectionTimeoutSeconds'),
             observeConfigValue(database, 'MobileEphemeralModeOfflinePersistenceTimerHours'),
+            observeConfigValue(database, 'MobileEphemeralModeAutoCacheCleanupDays'),
         ]).pipe(
             distinctUntilChanged(
-                ([prevEnabled, prevTimeout, prevPurgeHours], [nextEnabled, nextTimeout, nextPurgeHours]) =>
-                    prevEnabled === nextEnabled && prevTimeout === nextTimeout && prevPurgeHours === nextPurgeHours,
+                ([prevEnabled, prevTimeout, prevPurgeHours, prevCleanupDays],
+                    [nextEnabled, nextTimeout, nextPurgeHours, nextCleanupDays]) =>
```
Comment (larkox) [inline app/managers/ephemeral_mode_manager/index.ts]: Nitpick:

```suggestion
                (
                    [prevEnabled, prevTimeout, prevPurgeHours, prevCleanupDays],
                    [nextEnabled, nextTimeout, nextPurgeHours, nextCleanupDays],
                ) =>
```

File: app/managers/ephemeral_mode_manager/index.ts [hunk 24]
```diff
@@ -136,22 +149,28 @@ class EphemeralModeManagerSingleton {
         enabledStr: string | undefined,
         timeoutStr: string | undefined,
         purgeHoursStr: string | undefined,
+        cleanupDaysStr: string | undefined,
     ) => {
         const nextEnabled = enabledStr === 'true';
-        const nextThresholdMs = Math.max(0, Number(timeoutStr ?? '0')) * 1000;
-        const nextPurgeThresholdMs = Math.max(0, Number(purgeHoursStr ?? '0')) * 3600 * 1000;
+        const nextThresholdMs = parseNonNegativeConfigNumber(timeoutStr) * 1000;
+        const nextPurgeThresholdMs = parseNonNegativeConfigNumber(purgeHoursStr) * 3600 * 1000;
```
Comment (larkox) [inline app/managers/ephemeral_mode_manager/index.ts]: nitpick: We have a "toMilliseconds" (or something like that) function so it is clearer how much time we are waiting. So this could become...


```suggestion
        const nextPurgeThresholdMs =  toMillisecond({hour: parseNonNegativeConfigNumber(purgeHoursStr)});
```

And similar with the previous line too.

File: app/managers/ephemeral_mode_manager/index.test.ts [hunk 25]
```diff
@@ -328,6 +348,53 @@ describe('EphemeralModeManager', () => {
...
+            await advanceTimers(0);
+            expect(EphemeralModeManager.getAutoCacheCleanupDays(serverA)).toBe(7);
+
+            await updateConfig(serverA, {cleanupDays: 21});
+            await advanceTimers(0);
+
+            expect(EphemeralModeManager.getAutoCacheCleanupDays(serverA)).toBe(21);
+        });
+
+        it('falls back to 0 cleanup days when the server sends a non-numeric value', async () => {
```
Comment (larkox) [inline app/managers/ephemeral_mode_manager/index.test.ts:388]: Just to be thorough, having another tests with a negative value would be a nice to have.

File: app/actions/local/systems.ts [hunk 26]
```diff
@@ -358,7 +358,7 @@ export async function dataRetentionCleanPosts(serverUrl: string, postIds: string
         const batchSize = 1000;
         const deletePromises = [];
         for (let i = 0; i < postIds.length; i += batchSize) {
-            const batch = postIds.slice(i, batchSize);
+            const batch = postIds.slice(i, i + batchSize);
```
Comment (larkox) [inline app/actions/local/systems.ts:369]: Nice catch. Do we have thorough tests for this file? We should add a test for this case, just to be safe.

**Review by larkox — APPROVED** [review]
Several nitpicks, but nothing blocking.

That being said, I see several files with changes and no tests. I only mentioned one that should count for the coverage, but I wanted to raise it in case there is more that can be done there.

File: app/actions/local/ephemeral_mode/cleanup.test.ts [hunk 27]
```diff
@@ -0,0 +1,513 @@
...
+// cleanupDays=1 → cutoff = NOW - 1 day (DateConstants.SECONDS.DAY * 1000).
+const NOW = 100_000_000;
+const CUTOFF = NOW - (86_400 * 1000); // 13_600_000
+const OLD = 1_000_000; // below CUTOFF
+const RECENT = NOW; // above CUTOFF
+
+let database: Database;
+let operator: ServerDataOperator;
+
+async function writePiC(channelId: string, earliest: number, latest: number): Promise<PostsInChannelModel> {
```
Comment (enahum) [inline app/actions/local/ephemeral_mode/cleanup.test.ts]: wondering why this helpers instead of using the handlers from the operator

File: app/actions/local/ephemeral_mode/cleanup.test.ts [hunk 28]
```diff
@@ -0,0 +1,513 @@
...
+        const viewedChannelId = 'ch-viewed-few';
+        jest.spyOn(DatabaseManager, 'getActiveServerUrl').mockResolvedValue(SERVER_URL);
+        jest.mocked(NavigationStore.getScreensInStack).mockReturnValue([Screens.CHANNEL]);
+        jest.mocked(getCurrentChannelId).mockResolvedValue(viewedChannelId);
+        jest.mocked(EphemeralStore.getCurrentChannelOldestVisibleCreateAt).mockReturnValue(CUTOFF + 500);
+
+        // Only 5 posts older than anchor — fewer than BUFFER=20
+        await Promise.all(
+            Array.from({length: 5}, (_, idx) =>
+                writePost(`post-few-${idx + 1}`, viewedChannelId, (CUTOFF + 500) - ((idx + 1) * 10)),
```
Comment (enahum) [inline app/actions/local/ephemeral_mode/cleanup.test.ts:254]: with the operator handlers you definitely do not need to do this kind of things

File: app/actions/local/ephemeral_mode/cleanup.ts [hunk 29]
```diff
@@ -0,0 +1,294 @@
...
+    viewedChannelId: string | undefined;
+    viewedChannelLimit: number;
+    threadParentChannelId: string | undefined;
+    threadParentLimit: number;
+    viewedThreadId: string | undefined;
+    fileViewerPostId: string | undefined;
+    viewedPlaybookRunId: string | undefined;
+}
+
+async function getLastAutoCacheCleanupRun(database: Database): Promise<number> {
```
Comment (enahum) [inline app/actions/local/ephemeral_mode/cleanup.ts:42]: every action should have serverUrl and get the database from the DatabaseManager, not receive the Database directly, applies across the file

File: app/actions/local/post.ts [hunk 30]
```diff
@@ -390,21 +393,21 @@ export async function deletePosts(serverUrl: string, postIds: string[]) {
     try {
         const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
 
-        const postsFormatted = `'${postIds.join("','")}'`;
+        const placeholders = postIds.map(() => '?').join(',');
```
Comment (enahum) [inline app/actions/local/post.ts:398]: nice!!!

File: app/actions/local/systems.ts [hunk 31]
```diff
@@ -32,6 +32,7 @@ import {deletePostsForChannelsWithAutotranslation} from './channel';
 import {deletePosts} from './post';
 
 import type {DataRetentionPoliciesRequest} from '@actions/remote/systems';
+import { getFullErrorMessage } from '@utils/errors';
```
Comment (enahum) [inline app/actions/local/systems.ts]: weird the lint should complaint about this


```suggestion
import {getFullErrorMessage} from '@utils/errors';
```

File: app/actions/local/systems.ts [hunk 32]
```diff
@@ -358,7 +366,7 @@ export async function dataRetentionCleanPosts(serverUrl: string, postIds: string
         const batchSize = 1000;
         const deletePromises = [];
         for (let i = 0; i < postIds.length; i += batchSize) {
-            const batch = postIds.slice(i, batchSize);
+            const batch = postIds.slice(i, i + batchSize);
```
Comment (enahum) [inline app/actions/local/systems.ts:369]: nice!

File: app/screens/channel/channel_post_list/channel_post_list.tsx (hunk 21 above)
Comment (enahum) [reply app/screens/channel/channel_post_list/channel_post_list.tsx]: I think if you filter by postId and get the first element in the list you should be good

File: app/managers/ephemeral_mode_manager/index.ts (hunk 24 above)
Comment (enahum) [reply app/managers/ephemeral_mode_manager/index.ts]: agree on this

File: app/screens/pdf_viewer/pdf_viewer.tsx [hunk 33]
```diff
@@ -134,6 +137,25 @@ const PdfViewer = ({allowPdfLinkNavigation, fileId, filePath, siteURL}: Props) =
...
+        return () => {
+            active = false;
+            EphemeralStore.clearCurrentFileViewerPostId();
+        };
+    });
```
[the remark this answers is not in the document]
Comment (enahum) [reply app/screens/pdf_viewer/pdf_viewer.tsx:158]: can you check this comment, it looks correct

**Review by enahum — DISMISSED** [review]

File: app/actions/local/systems.ts (hunk 31 above)
Comment (carlisgg) [reply app/actions/local/systems.ts]: fixed

File: app/screens/channel/channel_post_list/channel_post_list.tsx (hunk 21 above)
Comment (carlisgg) [reply app/screens/channel/channel_post_list/channel_post_list.tsx]: I was not relying on query sorting by desc on createdAt, but I will do so and traverse in reverse adding a comment that this relies on the query sorting. I will keep the loop, it is more performant than a filter on an array.

File: app/products/playbooks/screens/playbook_run/playbook_run.tsx (hunk 22 above)
Comment (carlisgg) [reply app/products/playbooks/screens/playbook_run/playbook_run.tsx:210]: The `playbook-run id` is known up-front and we can then use it in `useDidMount` once to inform the playbook run instead rerun on empty and non empty emissions of the actual model.

File: app/queries/servers/post.ts
Comment (carlisgg) [reply app/queries/servers/post.ts:1]: I have added tests for the new function

**Review by edgarbellot — APPROVED** [review]
@carlisgg LGTM, thank you for addressing everything!

File: app/actions/local/ephemeral_mode/cleanup.ts (hunk 29 above)
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.ts:42]: the exported function already uses serverUrl and get the database from the manager, but for private unexported helper ones it is an established pattern to use the database directly as an argument like in `preparePostDeletion` in `posts.ts` actions. I have changed signatures for functions passing serverUrl and database though to just use one or the other.

File: app/actions/local/ephemeral_mode/cleanup.test.ts (hunk 27 above)
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.test.ts]: no handler produces PostsInChannel, it is actually derived from posts received. So the easiest for testing is just to directly write some fixtures. As fors `writePosts` the actual handler has some side effects, like touching in some cases the PiC table. It can be controlled via params but I don't want this test to rely on that so it is easier to write some posts in a controlled fashion like here. As for playbooks and ai threads you are right, so I will switch those to use the handlers.

File: app/actions/local/ephemeral_mode/cleanup.test.ts (hunk 28 above)
Comment (carlisgg) [reply app/actions/local/ephemeral_mode/cleanup.test.ts:254]: as mentioned above, not for writePosts.

**Comment by carlisgg** [comment]
@enahum waiting on you final approval here

**Review by enahum — APPROVED** [review]
@carlisgg sorry, seems that when I approved it did not marked it.. maybe a network error.. approving again.
